core/src/test/scala/unit/kafka/server/DescribeUserScramCredentialsRequestTest.scala

```diff
@@ -0,0 +1,128 @@
+ */
+class DescribeUserScramCredentialsRequestTest extends BaseRequestTest {
+  override def brokerPropertyOverrides(properties: Properties): Unit = {
+    properties.put(KafkaConfig.ControlledShutdownEnableProp, "false")
+    properties.put(KafkaConfig.AuthorizerClassNameProp, classOf[DescribeCredentialsTest.TestAuthorizer].getName)
+    properties.put(KafkaConfig.PrincipalBuilderClassProp, classOf[DescribeCredentialsTest.TestPrincipalBuilder].getName)
+  }
+
+  @Before
+  override def setUp(): Unit = {
+    DescribeCredentialsTest.principal = KafkaPrincipal.ANONYMOUS // default is to be authorized
+    super.setUp()
+  }
+
+  @Test
+  def testDescribeNothing(): Unit = {
+    val request = new DescribeUserScramCredentialsRequest.Builder(
+      new DescribeUserScramCredentialsRequestData()).build()
+    val response = sendDescribeUserScramCredentialsRequest(request)
+
+    val error = response.data.error
+    assertEquals("Expected no error when routed correctly", Errors.NONE.code, error)
+    assertEquals("Expected no credentials", 0, response.data.userScramCredentials.size)
+  }
+
+  @Test
+  def testDescribeNotController(): Unit = {
+    val request = new DescribeUserScramCredentialsRequest.Builder(
+      new DescribeUserScramCredentialsRequestData()).build()
+    val response = sendDescribeUserScramCredentialsRequest(request, notControllerSocketServer)
+
+    val error = response.data.error
+    assertEquals("Expected controller error when routed incorrectly", Errors.NOT_CONTROLLER.code, error)
+  }
+
+  @Test
+  def testDescribeNotAuthorized(): Unit = {
+    DescribeCredentialsTest.principal = DescribeCredentialsTest.UnauthorizedPrincipal
+
+    val request = new DescribeUserScramCredentialsRequest.Builder(
+      new DescribeUserScramCredentialsRequestData()).build()
+    val response = sendDescribeUserScramCredentialsRequest(request)
+
+    val error = response.data.error
+    assertEquals("Expected not authorized error", Errors.CLUSTER_AUTHORIZATION_FAILED.code, error)
+  }
+
+  @Test
+  def testDescribeSameUserTwice(): Unit = {
+    val user = new UserName().setName("user1")
+    val request = new DescribeUserScramCredentialsRequest.Builder(
+      new DescribeUserScramCredentialsRequestData().setUsers(List(user, user).asJava)).build()
+    val response = sendDescribeUserScramCredentialsRequest(request)
+
+    val error = response.data.error
+    assertEquals("Expected invalid request error", Errors.INVALID_REQUEST.code, error)
+  }
+
+
+  private def sendDescribeUserScramCredentialsRequest(request: DescribeUserScramCredentialsRequest, socketServer: SocketServer = controllerSocketServer): DescribeUserScramCredentialsResponse = {
+    connectAndReceive[DescribeUserScramCredentialsResponse](request, destination = socketServer)
+  }
+}
+
+object DescribeCredentialsTest {
+  val UnauthorizedPrincipal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "Unauthorized")
```

Comment:
It's kind of a code smell to be mutating static data during a JUnit test.  There are ways to run JUnit tests in parallel (although I think our current setup only runs at the granularity of test classes, not test methods?)  At the very least we'd have to document that this is not thread-safe and that it will prevent us from parellelizing this test class in the future.

Rather than doing this, how about having two separate test authorizers, and choosing the one you want for each individual JUnit test method?

In order to fit this into our bass-akwards inheritance based test design ( :( ) you could check the name of the test in `brokerPropertyOverrides` using one of these methods: https://stackoverflow.com/questions/473401/get-name-of-currently-executing-test-in-junit-4

Reply:
@cmccabe Thanks.  I copied this approach from `kafka.server.RequestQuotaTest`.  I fixed it here in `DesrcibeUserScramCredentialsRequestTest` and `AlterUserScramCredentialsRequestTest` but was unable to fix it in `RequestQuotaTest` because that test is checking every possible API, and of course certain APIs need to succeed for Kafka to start successfully.  So that test needs to mutate the behavior after `setUp()`.  These tests don't since these APIs aren't necessary for Kafka to start.  So I was able to fix it here but was unable to fix it where I got the approach from.